### PR TITLE
test: REST PHPUnit tests

### PR DESCRIPTION
**Pull request by Kallyan01**
## What

Added REST PHPUnit test cases.

## Why

Improves development workflow by enabling faster and more reliable validation of REST functionality, reducing repetitive manual testing efforts.

### Related Issue(s):

- https://github.com/rtCamp/OnePress/issues/71

### AI Disclosure

Hybrid approach = Written by Me + Copilot (Claude Opus 4.6, GPT 5.3) + Audited by me

## Testing Instructions

```bash
npm run wp-env:test -- start --xdebug=coverage    

npm run test:php -- tests/php/Unit/Modules/Rest
```

## Checklist

- [ ] I have read the [Contribution Guidelines](https://github.com/rtCamp/onesearch/blob/main/docs/CONTRIBUTING.md).
- [ ] I have read the [Development Guidelines](https://github.com/rtCamp/onesearch/blob/main/docs/DEVELOPMENT.md).
- [ ] I have added necessary tests to cover my changes.
- [ ] I have updated the project documentation as needed.
- [ ] My code has detailed inline documentation.
- [ ] My code is tested to the best of my abilities.
- [ ] My code passes all lints, tests, and checks.

<a href="https://playground.wordpress.net?blueprint-url=data:application/json,%7B%22%24schema%22%3A%22https%3A%2F%2Fplayground.wordpress.net%2Fblueprint-schema.json%22%2C%22meta%22%3A%7B%22title%22%3A%22OneSearch%20Demo%22%2C%22description%22%3A%22Cross-site%2C%20brand-aware%20search%20across%20a%20multi-brand%20WordPress%20network%2C%20powered%20by%20Algolia.%22%2C%22author%22%3A%22rtCamp%22%2C%22categories%22%3A%5B%22plugin%22%2C%22search%22%2C%22multisite%22%2C%22algolia%22%2C%22enterprise%22%5D%7D%2C%22landingPage%22%3A%22%2Fwp-admin%2Fplugins.php%22%2C%22preferredVersions%22%3A%7B%22php%22%3A%228.4%22%2C%22wp%22%3A%22latest%22%7D%2C%22features%22%3A%7B%22networking%22%3Atrue%7D%2C%22steps%22%3A%5B%7B%22step%22%3A%22login%22%2C%22username%22%3A%22admin%22%7D%2C%7B%22step%22%3A%22installPlugin%22%2C%22pluginData%22%3A%7B%22resource%22%3A%22url%22%2C%22url%22%3A%22https%3A%2F%2Fgithub.com%2FrtCamp%2Fonesearch%2Freleases%2Fdownload%2Fci-artifacts%2Fpr-212-332e02142210b58fa30b55fbd03fe509d34338a9.zip%22%7D%2C%22options%22%3A%7B%22activate%22%3Atrue%7D%7D%5D%7D" target="_blank" rel="noopener noreferrer">
  Open WordPress Playground Preview
</a>

### CODE REVIEW

File: tests/php/Unit/Modules/Rest/Governing_Data_Controller_GoverningSiteTest.php [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+	public function test_register_routes_for_governing_site(): void {
+		update_option( Settings::OPTION_SITE_TYPE, Settings::SITE_TYPE_GOVERNING );
+
+		$this->controller->register_routes();
+
+		$routes = rest_get_server()->get_routes();
+		$ns     = '/' . Governing_Data_Controller::NAMESPACE;
+
+		$this->assertArrayHasKey( $ns . '/brand-config', $routes );
+		$this->assertArrayHasKey( 'GET', $routes[ $ns . '/brand-config' ][0]['methods'] );
```
[the remark this answers is not in the document]
Comment (Kallyan01) [reply tests/php/Unit/Modules/Rest/Governing_Data_Controller_GoverningSiteTest.php:71]: Wrong

File: tests/php/Unit/Modules/Rest/Governing_Data_ControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+	public function test_register_routes_for_consumer_site(): void {
+		update_option( Settings::OPTION_SITE_TYPE, Settings::SITE_TYPE_CONSUMER );
+
+		$this->controller->register_routes();
+
+		$routes = rest_get_server()->get_routes();
+		$ns     = '/' . Governing_Data_Controller::NAMESPACE;
+
+		$this->assertArrayHasKey( $ns . '/brand-config', $routes );
+		$this->assertArrayHasKey( 'DELETE', $routes[ $ns . '/brand-config' ][0]['methods'] );
```
[the remark this answers is not in the document]
Comment (Kallyan01) [reply tests/php/Unit/Modules/Rest/Governing_Data_ControllerTest.php]: Wrong

File: tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+			$this->controller->test_is_url_from_host( 'https://example.com:8080', 'example.com', 9090 )
+		);
+	}
+
+	/**
+	 * URL without port defaults to 80.
+	 */
+	public function test_is_url_from_host_defaults_port_to_80(): void {
+		$this->assertTrue(
+			$this->controller->test_is_url_from_host( 'https://example.com', 'example.com', 80 )
```
[the remark this answers is not in the document]
Comment (Kallyan01) [reply tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php]: Port 80 is mentioned in the method is_url_from_host().

File: tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	}
+
+	/**
+	 * Expose the protected method for direct testing.
+	 *
+	 * @param string   $url  URL to check.
+	 * @param string   $host Host to compare.
+	 * @param int|null $port Optional port.
+	 */
+	public function test_is_url_from_host( string $url, string $host, ?int $port = null ): bool {
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php]: Is this function for testing or just grabbing the value from this function? If this is not the test function, then we can rename it to expose_is_url_from_host or something else

File: tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+			]
+		);
+
+		$response = $this->controller->get_shared_sites();
+		$data     = $response->get_data();
+
+		$this->assertTrue( $data['success'] );
+		$this->assertCount( 2, $data['shared_sites'] );
+		$this->assertSame( 'Brand A', $data['shared_sites'][0]['name'] );
+		$this->assertNotSame( 'Brand B', $data['shared_sites'][0]['name'] );
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php]: Does this assertion need to be?

File: tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+		$filter         = static function ( $preempt, $args, $url ) use ( &$requested_urls ) { // phpcs:ignore SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter
+			$requested_urls[] = $url;
+			return new \WP_Error( 'blocked', 'Intercepted' );
+		};
+		add_filter( 'pre_http_request', $filter, 10, 3 );
+
+		Governing_Data_Handler::clear_brand_config_cache();
+
+		remove_filter( 'pre_http_request', $filter );
+
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php:222]: This is fine, but we can also use try and finally blocks for handling this types of scenario.

File: tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+		);
+
+		$requested_urls = [];
+		$filter         = static function ( $preempt, $args, $url ) use ( &$requested_urls ) { // phpcs:ignore SlevomatCodingStandard.Functions.UnusedParameter.UnusedParameter
+			$requested_urls[] = $url;
+			return new \WP_Error( 'blocked', 'Intercepted' );
+		};
+		add_filter( 'pre_http_request', $filter, 10, 3 );
+
+		Governing_Data_Handler::clear_brand_config_cache( 'https://site-a.example.com/' );
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php:253]: Make sure the the valure stores with trainling slashes before doing the check.

File: tests/php/Unit/Modules/Rest/Search_ControllerTest.php [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+	// ── reindex ─────────────────────────────────────────────────────────
+
+	/**
+	 * Reindex on governing site without Algolia credentials returns failure response.
+	 */
+	public function test_reindex_standalone_returns_failure_without_algolia(): void {
+		// Governing with no Algolia: get_post_types_to_index() returns [] (no WP_Error),
+		// then index_all_posts() calls delete_by() which fails on missing Algolia
+		// credentials, collecting an error entry and returning success: false.
+		update_option( Settings::OPTION_SITE_TYPE, Settings::SITE_TYPE_GOVERNING );
```
Comment (sabbir1991) [inline tests/php/Unit/Modules/Rest/Search_ControllerTest.php]: You are setting the site type governing but the function name is standalone, do you think it can be `test_reindex_governing_retunrs_failure_without_algolia`

File: tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php (hunk 4 above)
Comment (Kallyan01) [reply tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php]: It's just for grabbing the value from the protected `is_url_from_host` method.

File: tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php (hunk 5 above)
Comment (Kallyan01) [reply tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php]: @sabbir1991 are you referring to this? 

```php
	$this->assertNotSame( 'Brand B', $data['shared_sites'][0]['name'] );
```

File: tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php (hunk 7 above)
Comment (Kallyan01) [reply tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php:253]: Could you please clarify this a little? Are you referring to the assert section?

File: tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php (hunk 5 above)
Comment (sabbir1991) [reply tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php]: Yes.

File: tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php (hunk 4 above)
Comment (sabbir1991) [reply tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php]: Then you can rename the function.

File: tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php (hunk 7 above)
Comment (sabbir1991) [reply tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php:253]: clear_brand_config_cache( 'https://site-a.example.com/' ) — called with trailing slash, but set_shared_sites stores 'https://site-a.example.com' (no trailing slash). If production does an exact string match, the filter never fires and assertCount( 1, $requested_urls ) will fail for the wrong reason. Normalize URLs to be consistent (both with or both without a trailing slash)

**Review by sabbir1991 — DISMISSED** [review]

File: tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+	/**
+	 * {@inheritDoc}
+	 */
+	public function set_up(): void {
+		parent::set_up();
+
+		$this->controller = new Concrete_REST_Controller();
+	}
+
+	// ── register_hooks ──────────────────────────────────────────────────
```
Comment (justlevine) [inline tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php]: Please get rid of all the AI slop 🙇

File: tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+	public function matches_host_url( string $url, string $host, ?int $port = null ): bool {
+		return $this->is_url_from_host( $url, $host, $port );
+	}
+}
+
+/**
+ * Tests for the abstract base REST controller.
+ */
+#[CoversClass( Abstract_REST_Controller::class )]
+class Abstract_REST_ControllerTest extends TestCase {
```
Comment (justlevine) [inline tests/php/Unit/Modules/Rest/Abstract_REST_ControllerTest.php]: We're making integration tests, the focus of our tests should be on outcomes, not implementations like the underlying Abstract classes.

Instead, we should add `#[CoversClass( Abstract_REST_Controller::class )]` to the top of the other `*_ControllerTest`s and only test here if there's particular edge cases that arent being covered.

(To fix, I recommend adding the #Covers*, delete all the the contents of `Abstract_REST_ControllerTest`, and then use the coverage report to see which of the existing tests here - if any - you should restore. )

File: tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+use OneSearch\Modules\Settings\Settings;
+use OneSearch\Tests\TestCase;
+use PHPUnit\Framework\Attributes\CoversClass;
+use WP_REST_Request;
+
+/**
+ * Tests for the basic options REST endpoints.
+ */
+#[CoversClass( Basic_Options_Controller::class )]
+class Basic_Options_ControllerTest extends TestCase {
```
Comment (justlevine) [inline tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php:24]: Per https://github.com/rtCamp/onesearch/pull/212#discussion_r3249270280
```suggestion
#[CoversClass( Abstract_Rest_Controller::class )]
#[CoversClass( Basic_Options_Controller::class )]
class Basic_Options_ControllerTest extends TestCase {
```

File: tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+	 * {@inheritDoc}
+	 */
+	public function set_up(): void {
+		parent::set_up();
+
+		global $wp_rest_server;
+		$wp_rest_server = null;
+
+		$this->controller = new Basic_Options_Controller();
+	}
```
Comment (justlevine) [inline tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php:57]: We're testing _outcomes_, so we want to make actual rest requests instead of hitting the controller.


```suggestion
	/**
	 * REST server.
	 */
	private ?\WP_Rest_Server $server;

	/**
	 * {@inheritDoc}
	 */
	public function set_up(): void {
		parent::set_up();

		global $wp_rest_server;
		$this->server = $wp_rest_server = new \WP_REST_Server;
  	do_action( 'rest_api_init' );
	}

	/**
	 * {@inheritDoc}
	 */
	public function tear_down(): void {
		global $wp_rest_server;
		$wp_rest_server = null;

		parent::tear_down();
	}
```

Then, for all of these test cases instead of trying to pass specific details to functions out of context, you would:

```php
	$request  = new WP_REST_Request( ... );
  $response = $this->server->dispatch( $request );
  $actual   = $response->get_data();

	$this->assert*();...
```

File: tests/php/Unit/Modules/Rest/Governing_Data_ControllerTest.php [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+
+	// ── Route registration ──────────────────────────────────────────────
+
+	/**
+	 * Governing site registers brand-config GET and all-post-types.
+	 *
+	 * @expectedIncorrectUsage register_rest_route
+	 */
+	public function test_register_routes_for_governing_site(): void {
+		update_option( Settings::OPTION_SITE_TYPE, Settings::SITE_TYPE_GOVERNING );
```
Comment (justlevine) [inline tests/php/Unit/Modules/Rest/Governing_Data_ControllerTest.php]: Only you're instantiating the endpoints for real per https://github.com/rtCamp/onesearch/pull/212#discussion_r3249417966 you'll need to move this to the set_up() too. You might need to break the file into separate `Governing_Data_Controller_{Governing|Brand}_SiteTest`

File: tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+use OneSearch\Modules\Rest\Governing_Data_Handler;
+use OneSearch\Modules\Settings\Settings;
+use OneSearch\Tests\TestCase;
+use PHPUnit\Framework\Attributes\CoversClass;
+
+/**
+ * Tests for brand config fetching, caching, and cross-site post type retrieval.
+ */
+#[CoversClass( Governing_Data_Handler::class )]
+class Governing_Data_HandlerTest extends TestCase {
```
Comment (justlevine) [inline tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php:21]: PTAL at the coverage report and use `add_filter( 'pre_http_request', ...)` to mock the `wp_safe_remote_request()` return value and finish testing those functions.

File: tests/php/Unit/Modules/Rest/Search_Controller_GoverningSiteTest.php
Comment (justlevine) [inline tests/php/Unit/Modules/Rest/Search_Controller_GoverningSiteTest.php:1]: We may need to mock some things to finish testing the important parts of this class. Will know more once everything's been converted to use `$this->server->dispatch()`

**Review by justlevine — CHANGES_REQUESTED** [review]
These test cases look really good - great job @Kallyan01 !

As noted, we want these to be outcome-focused integration tests and get as close to making real REST requests as possible instead of calling the controller methods directly, and gotta ditch the `// ai slop comments` everywhere. Once that's done it'l be easier to spot if there's something missing or more likely if there's some that are too "implementation-detail" focused that we want to ditch.

Either way, shouldn't take much before we can merge 🙌

File: tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function tear_down(): void {
+		global $wp_rest_server;
+		$wp_rest_server = null;
+
+		parent::tear_down();
+	}
```
[the remark this answers is not in the document]
Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Basic_Options_ControllerTest.php:57]: This isn't true. Settings are cleared in the teardown.

File: tests/php/Unit/Modules/Rest/Search_Controller_ConsumerSiteTest.php [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function tear_down(): void {
+		global $wp_rest_server;
+		$wp_rest_server = null;
+
+		parent::tear_down();
+	}
```
[the remark this answers is not in the document]
Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Search_Controller_ConsumerSiteTest.php:58]: they're torn down in the parent.

File: tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php [hunk 17]
```diff
@@ -0,0 +1,542 @@
...
+	/**
+	 * Returns error when site is not a consumer site.
+	 */
+	public function test_get_brand_config_returns_error_when_not_consumer_site(): void {
```
[the remark this answers is not in the document]
Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Governing_Data_HandlerTest.php:25]: they're torn down in the parent.

File: tests/php/Unit/Modules/Rest/Governing_Data_Controller_BrandSiteTest.php [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function tear_down(): void {
+		global $wp_rest_server;
+		$wp_rest_server = null;
+
+		parent::tear_down();
+	}
```
[the remark this answers is not in the document]
Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Governing_Data_Controller_BrandSiteTest.php:58]: these dont leak

File: tests/php/Unit/Modules/Rest/Governing_Data_Controller_StandaloneSiteTest.php [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function tear_down(): void {
+		global $wp_rest_server;
+		$wp_rest_server = null;
+
+		parent::tear_down();
+	}
```
[the remark this answers is not in the document]
Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Governing_Data_Controller_StandaloneSiteTest.php:60]: lies

File: tests/php/Unit/Modules/Rest/Governing_Data_Controller_GoverningSiteTest.php [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function tear_down(): void {
+		global $wp_rest_server;
+		$wp_rest_server = null;
+
+		parent::tear_down();
+	}
```
[the remark this answers is not in the document]
Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Governing_Data_Controller_GoverningSiteTest.php:61]: go away dumbot

File: tests/php/Unit/Modules/Rest/Search_Controller_GoverningSiteTest.php [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function tear_down(): void {
+		global $wp_rest_server;
+		$wp_rest_server = null;
+
+		parent::tear_down();
+	}
```
[the remark this answers is not in the document]
Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Search_Controller_GoverningSiteTest.php:61]: jfc nooooo

File: tests/php/Unit/Modules/Rest/Search_Controller_GoverningSiteTest.php
Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Search_Controller_GoverningSiteTest.php:1]: Okay, looks like the only thing here still missing is a successful `::update_algolia_credentials()`, but I think that's fine, considering the validation is done by the client we're about to replace.

Comment (justlevine) [reply tests/php/Unit/Modules/Rest/Search_Controller_GoverningSiteTest.php:1]: Same with `Abstract_REST_Controller`. We can come back for those later when we need them.

**Review by justlevine — APPROVED** [review]
